lib/kramdown/converter/math_engine/sskatex/details.rb

```diff
@@ -0,0 +1,264 @@
+    # change and the configuration remains unchanged.
+    JSCTX = {}
+
+    # ExecJS uses <tt>runtime = Runtimes.const_get(name)</tt> without checks.
+    # That is fragile and potentially insecure with arbitrary user input.
+    # Instead we use a fixed dictionary restricted to valid contents.
+    # Note that there are aliases like <tt>SpiderMonkey = Spidermonkey</tt>.
+    JSRUN_FROMSYM =
+      ExecJS::Runtimes.constants.inject({}) { |dict, name|
+        runtime = ExecJS::Runtimes.const_get(name)
+        dict[name] = runtime if runtime.is_a?(ExecJS::Runtime)
+        dict }
+
+    # This should really be provided by +ExecJS::Runtimes+:
+    # A map from JS Runtime classes to their respective symbols.
+    # (+.name+ is too verbose.)
+    # For aliases like <tt>SpiderMonkey = Spidermonkey</tt>,
+    # there are several possible choices for the symbol,
+    # but all are equivalent, so we do not care.
+    JSRUN_TOSYM = JSRUN_FROMSYM.invert
+
+    # Dictionary for escape sequences used in Javascript string literals
+    JS_ESCAPE = {
+      "\\" => "\\\\",
+      "\"" => "\\\"",
+      # Escaping single quotes not necessary in double-quoted string literals
+      #"'" => "\\'",
+      # JS does not recognize \a nor GNU's \e
+      # \b is ambiguous in regexps, as in Perl and Ruby
+      #"\b" => "\\b",
+      "\f" => "\\f",
+      "\n" => "\\n",
+      "\r" => "\\r",
+      "\t" => "\\t",
+      "\v" => "\\v",
+    }
+
+    # Turn a string into an equivalent Javascript literal, double-quoted.
+    # Similar to +.to_json+, but escape all non-ASCII codes as well.
+    def jsquote(str)
+      # No portable way of escaping Unicode points above 0xFFFF
+      '"%s"' % str.gsub(/[\0-\u{001F}""\\\u{0080}-\u{FFFF}]/u) { |c|
+        JS_ESCAPE[c] || "\\u%04x" % c.ord
+      }
+    end
+
+    # This should really be provided by +ExecJS::Runtimes+:
+    # A list of available JS engines, as symbols, in the order of preference.
+    def jsruns_available
+      ExecJS::Runtimes.runtimes.select(&:available?).map(&JSRUN_TOSYM)
+    end
+
+    # Given a +math_engine_opts+ dictionary,
+    # return a new JS engine context, initialized with the JS helper files,
+    # and with a JS object +katexopts+ containing general KaTeX options.
+    def newcontext(config)
+      # Configure JS engine. If specified, config[ :jsrun ] takes precedence,
+      # followed by an equivalent of ExecJS::Runtimes.autodetect,
+      # i.e. ENV_EXECJS_RUNTIME (parsed in a less adventurous way)
+      # or else the best available engine.
+      # If the result is invalid, unknown, or unavailable, errors will ensue.
+      jsrun = (config[:jsrun] ||
+               ENV_EXECJS_RUNTIME ||
+               JSRUN_TOSYM[ExecJS::Runtimes.best_available] ||
+               'Disabled').to_s.to_sym
+      ExecJS.runtime = JSRUN_FROMSYM[jsrun]
+      if config[:verbose]
+        warn "Available JS runtimes: #{jsruns_available.join(', ')}"
```

Comment:
Seems better (will try), unless such use implies actual *warning* semantics. The use by SsKaTeX is rather meant for *logging* under the `verbose` or `debug` options.

Reply:
Yeah, I get that. However, if one consciously enables verbose output it is okay to use the warning facility. This way the output will be correctly shown by whatever tool uses kramdown.

As for `debug`: That's different since it should only ever be used for development purposes. So using `warn` would be fine by me.

Reply:
I have now tried using `converter.warning` for the `verbose` option. For that, I had to reorganize SsKaTeX's private interfaces a bit (passing `converter` instead of just `config`). Unfortunately, a `jekyll build` now does not output those messages at all. It seems I have to go back to `warn` directly.
However, I'd keep the changed interfaces, just in case we need `converter.warning` for other things.